### PR TITLE
PV2 1845, 1849, 1850, 1851

### PR DESCRIPTION
**Pull request by tom-gough**
Code & acceptance test scenarios for:

PV2-1845
PV2-1849
PV2-1850
PV2-1851

### CODE REVIEW

File: src/SFA.DAS.Payments.AcceptanceTests.Core/Automation/DcHelper.cs [hunk 1]
```diff
@@ -210,6 +214,35 @@ public async Task SendIlrSubmission(List<FM36Learner> learners, long ukprn, shor
...
+        public int GetPaymentsCount(long ukprn, CollectionPeriod collectionPeriod)
+        {
+            return dataContext.Payment.Count(x => x.Ukprn == ukprn && x.CollectionPeriod.Period == collectionPeriod.Period && x.CollectionPeriod.AcademicYear == collectionPeriod.AcademicYear);
+        }
+
+        public int GetRequiredPaymentsCount(long ukprn, CollectionPeriod collectionPeriod)
+        {
+            return dataContext.RequiredPaymentEvent.Count(x => x.Ukprn == ukprn && x.CollectionPeriod.Period == collectionPeriod.Period && x.CollectionPeriod.AcademicYear == collectionPeriod.AcademicYear);
+        }
+
```
Comment (carlbm) [inline src/SFA.DAS.Payments.AcceptanceTests.Core/Automation/DcHelper.cs:245]: Can you move these to another helper (as discussed)

File: src/SFA.DAS.Payments.EarningEvents.Application.UnitTests/ApprenticeshipContractTypeEarningsEventBuilderTest.cs [hunk 2]
```diff
@@ -728,6 +735,23 @@ public void CreatedTheCorrectNumberOfEarningEventsWhenLearningSupportIncluded()
             events.Single().IncentiveEarnings.Single(x => x.Type == IncentiveEarningType.LearningSupport).Periods.Should().HaveCount(12);
         }
 
+
+        [Test]
```
Comment (carlbm) [inline src/SFA.DAS.Payments.EarningEvents.Application.UnitTests/ApprenticeshipContractTypeEarningsEventBuilderTest.cs:739]: Can you add some tests for the boundaries? Redundancy date after learning, before start of learning etc?

File: src/SFA.DAS.Payments.EarningEvents.Application.UnitTests/Mapping/RedundancyEarningEventFactoryTest.cs [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+        public void CreateRedundancyContractType_CreatesCorrectContractTypeEvents(Type inputType, Type expectedType)
+        {
+            var factory = new RedundancyEarningEventFactory(Mapper.Instance);
+            ApprenticeshipContractTypeEarningsEvent earningEvent;
+
+            earningEvent =
+                factory.CreateRedundancyContractTypeEarningsEvent(
+                    (ApprenticeshipContractTypeEarningsEvent) Activator.CreateInstance(inputType));
+
+            Assert.AreEqual(expectedType, earningEvent.GetType());
```
Comment (carlbm) [inline src/SFA.DAS.Payments.EarningEvents.Application.UnitTests/Mapping/RedundancyEarningEventFactoryTest.cs:38]: Any reason not to use fluent assertions?

File: src/SFA.DAS.Payments.EarningEvents.Application/Services/RedundancyEarningService.cs [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+
+            if (month < 8)
+            {
+                period = (byte) (month + 5);
+            }
+            else
+            {
+                period = (byte) (month - 7);
+            }
+            return period;
```
Comment (carlbm) [inline src/SFA.DAS.Payments.EarningEvents.Application/Services/RedundancyEarningService.cs:118]: This code exists in the tests too. Can you put it somewhere available to both? (I think it may even exist somewhere else too!)

File: src/SFA.DAS.Payments.RequiredPayments.Application/Interfaces/IEarningEventProcessor.cs [hunk 5]
```diff
@@ -21,10 +21,16 @@ CancellationToken cancellationToken
     }
 
 
-    public interface IApprenticeshipContractType2EarningsEventProcessor : IEarningEventProcessor<ApprenticeshipContractType2EarningEvent>
```
Comment (carlbm) [inline src/SFA.DAS.Payments.RequiredPayments.Application/Interfaces/IEarningEventProcessor.cs:24]: Was this change intended? Seems outside of redundancy?

File: src/SFA.DAS.Payments.EarningEvents.Application/Services/RedundancyEarningService.cs [hunk 6]
```diff
@@ -0,0 +1,121 @@
...
+
+        private void RemovePreRedundancyEarningPeriods(Earning earning, byte redundancyPeriod)
+        {
+            var periods = earning.Periods.ToList();
+            periods.RemoveAll(p => p.Period < redundancyPeriod);
+            earning.Periods = periods.AsReadOnly();
+        }
+
+        public List<FunctionalSkillEarningsEvent> SplitFunctionSkillEarningByRedundancyDate(FunctionalSkillEarningsEvent functionalSkillEarning,
+            DateTime priceEpisodeRedStartDate)
```
Comment (hardik-desai-kcl) [inline src/SFA.DAS.Payments.EarningEvents.Application/Services/RedundancyEarningService.cs:70]: as discussed it might be easier to understand code if you keep the pattern same as List<ApprenticeshipContractTypeEarningsEvent> SplitContractEarningByRedundancyDate, i.e. sub functions and I can see some code reuse potential as well, i.e. change the input param to earning.Periods in RemoveRedundancyEarningPeriods and RemovePreRedundancyEarningPeriods

File: src/SFA.DAS.Payments.EarningEvents.Application.UnitTests/ApprenticeshipContractTypeEarningsEventBuilderTest.cs [hunk 7]
```diff
@@ -728,6 +735,23 @@ public void CreatedTheCorrectNumberOfEarningEventsWhenLearningSupportIncluded()
...
+        {
+            var builder = new ApprenticeshipContractTypeEarningsEventBuilder(
+                new ApprenticeshipContractTypeEarningsEventFactory(), redundancyEarningService.Object,
+                mapper);
+            var processLearnerCommand = FileHelpers.CreateFromFile(filename, learnerRefNo);
+
+            redundancyEarningService.Setup(x => x.SplitContractEarningByRedundancyDate(It.IsAny<ApprenticeshipContractType2EarningEvent>(), It.IsAny<DateTime>()))
+                .Returns(new List<ApprenticeshipContractTypeEarningsEvent>());
+
+            builder.Build(processLearnerCommand);
```
[the remark this answers is not in the document]
Comment (keenesolutions) [reply src/SFA.DAS.Payments.EarningEvents.Application.UnitTests/ApprenticeshipContractTypeEarningsEventBuilderTest.cs:750]: This is covered by RedundancyEarningServiceTests

File: src/SFA.DAS.Payments.EarningEvents.Application.UnitTests/Mapping/RedundancyEarningEventFactoryTest.cs [hunk 8]
```diff
@@ -0,0 +1,98 @@
...
+        [TestCase(typeof(ApprenticeshipContractType1EarningEvent),
+            typeof(ApprenticeshipContractType1RedundancyEarningEvent))]
+        [TestCase(typeof(ApprenticeshipContractType2EarningEvent),
+            typeof(ApprenticeshipContractType2RedundancyEarningEvent))]
+        public void CreateRedundancyContractType_CreatesCorrectContractTypeEvents(Type inputType, Type expectedType)
+        {
+            var factory = new RedundancyEarningEventFactory(Mapper.Instance);
+            ApprenticeshipContractTypeEarningsEvent earningEvent;
+
+            earningEvent =
```
[the remark this answers is not in the document]
Comment (keenesolutions) [reply src/SFA.DAS.Payments.EarningEvents.Application.UnitTests/Mapping/RedundancyEarningEventFactoryTest.cs:34]: done

File: src/SFA.DAS.Payments.EarningEvents.Application.UnitTests/Mapping/RedundancyEarningEventFactoryTest.cs (hunk 3 above)
Comment (keenesolutions) [reply src/SFA.DAS.Payments.EarningEvents.Application.UnitTests/Mapping/RedundancyEarningEventFactoryTest.cs:38]: Nope... Changed to use fluent assertions

File: src/SFA.DAS.Payments.EarningEvents.Application/Mapping/RedundancyEarningEventFactory.cs [hunk 9]
```diff
@@ -0,0 +1,41 @@
...
+        public ApprenticeshipContractTypeEarningsEvent CreateRedundancyContractTypeEarningsEvent(
+            ApprenticeshipContractTypeEarningsEvent earningEvent)
+        {
+            if (earningEvent is ApprenticeshipContractType1EarningEvent act1)
+                return mapper.Map<ApprenticeshipContractType1RedundancyEarningEvent>(act1);
+
+            if (earningEvent is ApprenticeshipContractType2EarningEvent act2)
+                return mapper.Map<ApprenticeshipContractType2RedundancyEarningEvent>(act2);
+
+            return null;
```
[the remark this answers is not in the document]
Comment (keenesolutions) [reply src/SFA.DAS.Payments.EarningEvents.Application/Mapping/RedundancyEarningEventFactory.cs:26]: There is a type ApprenticeshipContractTypeNoneEarningEvent which why I returned null here.

File: src/SFA.DAS.Payments.EarningEvents.Application/Services/RedundancyEarningService.cs (hunk 6 above)
Comment (keenesolutions) [reply src/SFA.DAS.Payments.EarningEvents.Application/Services/RedundancyEarningService.cs:70]: Done.

File: src/SFA.DAS.Payments.EarningEvents.Application/Services/RedundancyEarningService.cs (hunk 4 above)
Comment (keenesolutions) [reply src/SFA.DAS.Payments.EarningEvents.Application/Services/RedundancyEarningService.cs:118]: Moved to SFA.DAS.Payments.Core

File: src/SFA.DAS.Payments.RequiredPayments.Application/Interfaces/IEarningEventProcessor.cs (hunk 5 above)
Comment (keenesolutions) [reply src/SFA.DAS.Payments.RequiredPayments.Application/Interfaces/IEarningEventProcessor.cs:24]: Loosened contract on the processor so that it can process both reduncany and non redunancy ACT2 events

File: src/SFA.DAS.Payments.EarningEvents.Application.UnitTests/ApprenticeshipContractTypeEarningsEventBuilderTest.cs (hunk 2 above)
Comment (keenesolutions) [reply src/SFA.DAS.Payments.EarningEvents.Application.UnitTests/ApprenticeshipContractTypeEarningsEventBuilderTest.cs:739]: This will require a new FM36 to cover the scenario.

File: src/SFA.DAS.Payments.EarningEvents.Application.UnitTests/ApprenticeshipContractTypeEarningsEventBuilderTest.cs (hunk 7 above)
[the remark this answers is not in the document]
Comment (keenesolutions) [reply src/SFA.DAS.Payments.EarningEvents.Application.UnitTests/ApprenticeshipContractTypeEarningsEventBuilderTest.cs:750]: All I'm trying to cover here is the fact that the RedundancyEarningService method was invoked.

File: src/SFA.DAS.Payments.EarningEvents.Application/Services/RedundancyEarningService.cs (hunk 6 above)
Comment (hardik-desai-kcl) [reply src/SFA.DAS.Payments.EarningEvents.Application/Services/RedundancyEarningService.cs:72]: This is what I meant 

        ```
public List<ApprenticeshipContractTypeEarningsEvent> SplitContractEarningByRedundancyDate(ApprenticeshipContractTypeEarningsEvent earningEvent, DateTime redundancyDate)
        {
            var splitResults = new List<ApprenticeshipContractTypeEarningsEvent>();

            var redundancyPeriod = redundancyDate.GetPeriodFromDate();

            var redundancyEarningEvent = redundancyEarningEventFactory.CreateRedundancyContractTypeEarningsEvent(earningEvent);

            earningEvent.OnProgrammeEarnings.ForEach(earning => { earning.Periods = RemoveRedundancyPeriods(earning.Periods, redundancyPeriod); });
            earningEvent.IncentiveEarnings.ForEach(earning => { earning.Periods = RemoveRedundancyPeriods(earning.Periods, redundancyPeriod); });

            redundancyEarningEvent.OnProgrammeEarnings.ForEach(earning =>
                                                               {
                                                                   earning.Periods = RemovePreRedundancyPeriods(earning.Periods, redundancyPeriod);
                                                                   SetRedundancyPeriodsToFullContribution(earning.Periods);
                                                               });

            redundancyEarningEvent.IncentiveEarnings.ForEach(earning =>
                                                             {
                                                                 earning.Periods = RemovePreRedundancyPeriods(earning.Periods, redundancyPeriod);
                                                                 SetRedundancyPeriodsToFullContribution(earning.Periods);
                                                             });

            splitResults.Add(earningEvent);
            splitResults.Add(redundancyEarningEvent);

            return splitResults;
        }

        public List<FunctionalSkillEarningsEvent> SplitFunctionSkillEarningByRedundancyDate(FunctionalSkillEarningsEvent functionalSkillEarning, DateTime priceEpisodeRedStartDate)
        {
            var splitResults = new List<FunctionalSkillEarningsEvent>();
            var redundancyPeriod = GetPeriodFromDate(priceEpisodeRedStartDate);

            var redundancyEarningEvent = redundancyEarningEventFactory.CreateRedundancyFunctionalSkillTypeEarningsEvent(functionalSkillEarning);

            foreach (var earning in functionalSkillEarning.Earnings)
            {
                earning.Periods = RemoveRedundancyPeriods(earning.Periods, redundancyPeriod);
            }

            foreach (var earning in redundancyEarningEvent.Earnings)
            {
                earning.Periods = RemovePreRedundancyPeriods(earning.Periods, redundancyPeriod);
                SetRedundancyPeriodsToFullContribution(earning.Periods);
            }

            splitResults.Add(functionalSkillEarning);
            splitResults.Add(redundancyEarningEvent);

            return splitResults;
        }

        private static void SetRedundancyPeriodsToFullContribution(IEnumerable<EarningPeriod> periods)
        {
            foreach (var period in periods)
            {
                period.SfaContributionPercentage = 1m;
            }
        }

        private static ReadOnlyCollection<EarningPeriod> RemovePreRedundancyPeriods(IEnumerable<EarningPeriod> periods, byte redundancyPeriod)
        {
            var allPeriods = periods.ToList();
            allPeriods.RemoveAll(p => p.Period < redundancyPeriod);
            return allPeriods.AsReadOnly();
        }

        private static ReadOnlyCollection<EarningPeriod> RemoveRedundancyPeriods(IEnumerable<EarningPeriod> periods, byte redundancyPeriod)
        {
            var allPeriods = periods.ToList();
            allPeriods.RemoveAll(p => p.Period >= redundancyPeriod);
            return allPeriods.AsReadOnly();
        }
```

Comment (hardik-desai-kcl) [reply src/SFA.DAS.Payments.EarningEvents.Application/Services/RedundancyEarningService.cs:72]: This is what I meant

`        public List<ApprenticeshipContractTypeEarningsEvent> SplitContractEarningByRedundancyDate(ApprenticeshipContractTypeEarningsEvent earningEvent, DateTime redundancyDate)
        {
            var splitResults = new List<ApprenticeshipContractTypeEarningsEvent>();

            var redundancyPeriod = redundancyDate.GetPeriodFromDate();

            var redundancyEarningEvent = redundancyEarningEventFactory.CreateRedundancyContractTypeEarningsEvent(earningEvent);

            earningEvent.OnProgrammeEarnings.ForEach(earning => { earning.Periods = RemoveRedundancyPeriods(earning.Periods, redundancyPeriod); });
            earningEvent.IncentiveEarnings.ForEach(earning => { earning.Periods = RemoveRedundancyPeriods(earning.Periods, redundancyPeriod); });

            redundancyEarningEvent.OnProgrammeEarnings.ForEach(earning =>
                                                               {
                                                                   earning.Periods = RemovePreRedundancyPeriods(earning.Periods, redundancyPeriod);
                                                                   SetRedundancyPeriodsToFullContribution(earning.Periods);
                                                               });

            redundancyEarningEvent.IncentiveEarnings.ForEach(earning =>
                                                             {
                                                                 earning.Periods = RemovePreRedundancyPeriods(earning.Periods, redundancyPeriod);
                                                                 SetRedundancyPeriodsToFullContribution(earning.Periods);
                                                             });

            splitResults.Add(earningEvent);
            splitResults.Add(redundancyEarningEvent);

            return splitResults;
        }

        public List<FunctionalSkillEarningsEvent> SplitFunctionSkillEarningByRedundancyDate(FunctionalSkillEarningsEvent functionalSkillEarning, DateTime priceEpisodeRedStartDate)
        {
            var splitResults = new List<FunctionalSkillEarningsEvent>();
            var redundancyPeriod = GetPeriodFromDate(priceEpisodeRedStartDate);

            var redundancyEarningEvent = redundancyEarningEventFactory.CreateRedundancyFunctionalSkillTypeEarningsEvent(functionalSkillEarning);

            foreach (var earning in functionalSkillEarning.Earnings)
            {
                earning.Periods = RemoveRedundancyPeriods(earning.Periods, redundancyPeriod);
            }

            foreach (var earning in redundancyEarningEvent.Earnings)
            {
                earning.Periods = RemovePreRedundancyPeriods(earning.Periods, redundancyPeriod);
                SetRedundancyPeriodsToFullContribution(earning.Periods);
            }

            splitResults.Add(functionalSkillEarning);
            splitResults.Add(redundancyEarningEvent);

            return splitResults;
        }

        private static void SetRedundancyPeriodsToFullContribution(IEnumerable<EarningPeriod> periods)
        {
            foreach (var period in periods)
            {
                period.SfaContributionPercentage = 1m;
            }
        }

        private static ReadOnlyCollection<EarningPeriod> RemovePreRedundancyPeriods(IEnumerable<EarningPeriod> periods, byte redundancyPeriod)
        {
            var allPeriods = periods.ToList();
            allPeriods.RemoveAll(p => p.Period < redundancyPeriod);
            return allPeriods.AsReadOnly();
        }

        private static ReadOnlyCollection<EarningPeriod> RemoveRedundancyPeriods(IEnumerable<EarningPeriod> periods, byte redundancyPeriod)
        {
            var allPeriods = periods.ToList();
            allPeriods.RemoveAll(p => p.Period >= redundancyPeriod);
            return allPeriods.AsReadOnly();
        }`

File: src/SFA.DAS.Payments.AcceptanceTests.EndToEnd/FM36TestFiles/PV2-1845-Levy-Learner-Made-Redundant-In-Last-6-Months-R07.json [hunk 10]
```diff
@@ -0,0 +1,1771 @@
+{
```
[the remark this answers is not in the document]
Comment (tom-gough) [reply src/SFA.DAS.Payments.AcceptanceTests.EndToEnd/FM36TestFiles/PV2-1845-Levy-Learner-Made-Redundant-In-Last-6-Months-R07.json:1]: When we originally discussed using FM36 files for the acceptance tests in the dev/test session we agreed not to modify them away from how they are generated (beyond anonymisation of ULN/UKPRN/Learner Ref) for fear of creating un-realistic tests.
